### PR TITLE
fix quoting strings challenge

### LINKED ISSUE #12865: [beta] bug in "Quoting String with Single Quotes" challenge
Challenge [quoting-strings-with-single-quotes](http://beta.freecodecamp.com/en/challenges/basic-javascript/quoting-strings-with-single-quotes) has an issue.
User Agent is: <code>Mozilla/5.0 (X11; CrOS x86_64 8872.76.0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.105 Safari/537.36</code>.
Please describe how to reproduce this issue, and include links to screenshots if possible.

```js


var myStr = '<a href="http://www.example.com" target="_blank">Link</a>';

```

### PR DESCRIPTION
**Pull request by no-stack-dub-sack**
#### Pre-Submission Checklist

- [x] Your pull request targets the `staging` branch of freeCodeCamp.
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [x] All new and existing tests pass the command `npm test`. Use `git commit --amend` to amend any fixes.

#### Type of Change

- [x] Small bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)

#### Checklist:

- [x] Tested changes locally.
- [x] Closes currently open issue: Closes #12865 

#### Description

The [Quoting Strings with Single Quotes](http://beta.freecodecamp.com/en/challenges/basic-javascript/quoting-strings-with-single-quotes) challenge was affected when `'use-strict'` was added to the comment that is invisible to users. Similar to the issue seen in #12722. This challenge uses a regex to count the quotes in the `code` variable, so `'use strict'` and the `tail` was throwing off the count. Logging `code` looks like this:
```
<script>

;/*fcc*/
'use strict';

var myStr = '<a href="http://www.example.com" target="_blank">Link</a>;';
;/*fcc*/

(function() { return "myStr = " + myStr; })();

;/*fcc*/
</script>
<!--fcc-->
```

So there is an additional pair of both `'` and `"`. I tried to just accommodate for this by changing the test to account for the extra quotes, but my guess is that since this comment is not actually generated in nodes test environment, approaching it this way caused running `npm test` to fail.

So instead, I created an IIFE that executes within the assert and returns `true` or `false`. It slices `code` down to only the relevant parts then just runs the original regex matches against the new string.

I also added `blockquote`s for the examples to be consistent with other challenges and added `Do not change anything else.` to the instructions because we are also testing that the string remains unchanged in the first test but don't explicitly say it anywhere, so if someone does what the tests ask but changes the URL in the string, the tests will fail with an erroneous error message.

### CODE REVIEW

File: seed/challenges/02-javascript-algorithms-and-data-structures/basic-javascript.json [hunk 1]
```diff
@@ -1133,8 +1133,9 @@
       ],
       "tests": [
         "assert(!/\\\\/g.test(code) && myStr === '<a href=\"http://www.example.com\" target=\"_blank\">Link</a>', 'message: Remove all the <code>backslashes</code> (<code>\\</code>)');",
-        "assert(code.match(/\"/g).length === 4 && code.match(/'/g).length === 2, 'message: You should have two single quotes <code>&#39;</code> and four double quotes <code>&quot;</code>');"
+        "assert((function() { const strict = \"'use strict';\"; const tail = '(function() { return \"myStr = \" + myStr; })();'; const sliced = code.slice(code.indexOf(strict) + strict.length, code.indexOf(tail)); if (sliced.match(/\"/g).length === 4 && sliced.match(/'/g).length === 2) return true; return false; })() === true, 'message: You should have two single quotes <code>&#39;</code> and four double quotes <code>&quot;</code>');"
```
Comment (Greenheart) [inline seed/challenges/02-javascript-algorithms-and-data-structures/basic-javascript.json]: This solution assumes that the tail in the seed-file matches the tail that right now is hardcoded in the test. It could work, but it could also cause problems if we're unlucky.

A possible solution could be to just use the start of the tail section as argument for `code.indexOf(tail)`. If we added the comment `/*Required for test. Do not remove.*/` to the beginning of the tail `IIFE`, we could use this "constant string" to decide where the tail starts without having to rely on the exact code of the tail section.

Comment (Greenheart) [inline seed/challenges/02-javascript-algorithms-and-data-structures/basic-javascript.json]: We also have to consider `const` in the test `IIFE`. Can we safely assume that all environments that will run this challenge can handle ES6? 

Most semi-modern browsers can, but I'm not sure we want to break backwards compatibility with older browsers.

Even though it hurts (Gah, `var`!?), I suggest we replace `const` with `var`.

**Review by Greenheart — CHANGES_REQUESTED** [review]
Nice solution with the IIFE!

Looking forward to your thoughts :blush:

File: seed/challenges/02-javascript-algorithms-and-data-structures/basic-javascript.json (hunk 1 above)
Comment (no-stack-dub-sack) [inline seed/challenges/02-javascript-algorithms-and-data-structures/basic-javascript.json]: @Greenheart good call, I'll make a quick update

Comment (no-stack-dub-sack) [inline seed/challenges/02-javascript-algorithms-and-data-structures/basic-javascript.json]: @Greenheart good idea - yeah I thought of this that the test would need to be changed if the tail ever changes, so this sounds like a safer bet, I should've thought of this myself 😄

**Comment by no-stack-dub-sack** [comment]
@Greenheart Awesome! Thanks for your review and excellent suggestions (I left some comments in the outdated above). This should be all set now!

**Review by systimotic — COMMENTED** [review]
Previously, to fix head or tail issues, we have modified the test to look for a different amount.
While I think this solution is really cool, and probably more robust, I think it would be simpler to change the amount of quotes we're looking for. If the head or tail are ever modified to include a different amount of quotes, the CI should pick up on it.

**Review by lebanon777 — APPROVED** [review]

**Comment by Greenheart** [comment]
@systimotic Do you mean to `assert(numberOfQuotes === fccHeadAndTailQuotes + expectedSolutionQuotes)` instead of only counting what's in the camper's code?

**Comment by systimotic** [comment]
@Greenheart Yep. I prefer changing the amount for readability. Although this solution is really neat, it's quite hard to follow what's going on.

**Comment by Greenheart** [comment]
@systimotic I agree, it's easier to understand and update tests that use some numbers.

@no-stack-dub-sack @systimotic Should we go with the counting solution instead?

I think that would mean a test case looking something like this (except the exact numbers):
`assert((function(){ var fccHeadAndTailQuotes = X, expectedSolutionQuotes = Y, actualQuotes = code.match(/['"]/g).length; return actualQuotes === fccHeadAndTailQuotes + expectedSolutionQuotes; }()), 'message: You should have two single quotes <code>&#39;</code> and four double quotes <code>&quot;</code>');`

**Comment by systimotic** [comment]
@Greenheart I believe we only need to change `.length === 2 ` to `length === 4`

**Comment by Greenheart** [comment]
@systimotic Right!

Then this would be our new test case:
`assert(/* the following numbers include quotes from challenge head & tail */ code.match(/'/g).length === 4 && code.match(/'/g).length === 4, 'message: You should have two single quotes <code>&#39;</code> and four double quotes <code>&quot;</code>');`

@no-stack-dub-sack This also means we no longer need the IIFE or the "constant" comments.

**Comment by no-stack-dub-sack** [comment]
@Greenheart @systimotic So, naturally, this was my first instinct, and is clearly a much simpler solution, however, as I noted when I originally made the PR, I couldn't get `npm test` to pass with this solution, hence the fancy IIFE. Is this something that is OK for a change like this (not passing `npm test`) and this would be repaired during the CI check or something? Not too sure how all the CI stuff works, but that's the reason I didn't submit the PR with that solution

**Comment by no-stack-dub-sack** [comment]
@Greenheart Ok, I spoke with @systimotic and figured out that the original solution I had tried and that he suggested (just adjusting the count) will work and `npm test` will pass, as long as either the `solution` JSON is removed, or the `head` and `tail` are added to it. So to make it clearer for anyone looking at this later, I added the head and tail to the solution so you can see what we are counting. Should be good to merge now if you all agree that this is ok

**Review by Greenheart — APPROVED** [review]
@no-stack-dub-sack @systimotic Thanks for fixing this! :smile:
